source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -18,17 +21,115 @@ MapdataCheckPoint::MapdataCheckPoint(const SData *data)
+    m_id = id;
+    auto checkPathAccessor = CourseMap::Instance()->getCheckPathAccessor();
+
+    // Calculate the quadrilateral's `m_prevPoints`. If the check point is the first in its group,
+    // it has multiple previous checkpoints defined by its preceding checkpaths
+    if (m_prevPt == -1) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        m_prevCount = 0;
+
+        for (auto [i, prevID] : std::views::enumerate(checkpath->prev())) {
+            if (prevID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *prev = checkPathAccessor->get(prevID);
+            m_prevPoints[i] = accessor.get(prev->end());
+            m_prevCount++;
+        }
+    } else {
+        m_prevPoints[0] = accessor.get(m_prevPt);
+        m_prevCount++;
+    }
+
+    // Calculate the quadrilateral's `m_nextPoints`. If the checkpoint is the last in its group, it
+    // can have multiple quadrilaterals (and nextCheckpoint) which are determined by its next
+    // path(s)
+    if (m_nextPt == -1) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        m_nextCount = 0;
+
+        for (auto [i, nextID] : std::views::enumerate(checkpath->next())) {
+            if (nextID == 0xff) {
+                continue;
+            }
+
+            MapdataCheckPath *next = checkPathAccessor->get(nextID);
+            m_nextPoints[i].checkpoint = accessor.get(next->start());
+            m_nextCount++;
+        }
+    } else {
+        m_nextPoints[0].checkpoint = accessor.get(m_nextPt);
+        m_nextCount++;
+    }
+
+    // Form the checkpoint's quadrilateral(s)
+    for (size_t i = 0; i < 6; i++) {
+        if (i < m_nextCount) {
+            auto &nextLinked = m_nextPoints[i];
+            auto *nextPoint = nextLinked.checkpoint;
+
+            nextLinked.distance = (nextPoint->m_midpoint - m_midpoint).normalise();
+            nextLinked.p0diff =
+                    EGG::Vector2f(nextPoint->m_left.x - m_left.x, nextPoint->m_left.y - m_left.y);
+            nextLinked.p1diff = EGG::Vector2f(nextPoint->m_right.x - m_right.x,
+                    nextPoint->m_right.y - m_right.y);
+        }
+    }
+}
+
+s8 MapdataCheckPoint::checkArea() const {
+    return m_checkArea;
+}
+
+MapdataCheckPoint *MapdataCheckPoint::nextPoint(size_t i) const {
+    ASSERT(i < m_nextPoints.size());
+    return m_nextPoints[i].checkpoint;
+}
+
+/// @brief Finds indices of the finish line and last key checkpoint; also calls initCheckpointLinks
+/// for all checkpoints
+/// @addr{0x80515244}
+void MapdataCheckPointAccessor::init() {
+    s8 lastKcpType = -1;
+    s16 finishLineCheckpointId = -1;
+
+    for (size_t ckptId = 0; ckptId < size(); ckptId++) {
+        MapdataCheckPoint *checkpoint = get(ckptId);
+        s8 checkArea = checkpoint->checkArea();
+        checkpoint->initCheckpointLinks(*this, ckptId);
+        checkpoint = get(ckptId);
```

Comment:
At any given iteration, `ckptId` is the same as it was on line 111. This means that `checkpoint` (I think) is identical between lines 111 and 114, which means you should just delete line 114,